### ISSUE TITLE
Running Puppeteer with Firefox in AWS lambda fails due to missing write permissions for the Firefox profile

### ISSUE THREAD

**Issue by telltokichu** [puppeteer/puppeteer#13086]
I am running puppeteer in aws lambda with executablePath : await chromium.executablePath() which is from @sparticuz/chromium package.It runs flawlessly but when i want to run firefox with puppeteer with product: "firefox" it works fine in local environment but in aws lambda it fails.
Is  there any way to run puppeteer firefox in aws lambda with firefox binary or with npm package similar to @sparticuz/chromium ?
puppeteer version - 21.6.1

**Comment by OrKoN** [comment]
@whimboo do you know about any documentation for running Firefox in AWS Lambda?

In general, this is out of scope for Puppeteer to make sure browsers can run on Lambda. The system should be able to run the browser in order to use Puppeteer and Lambda environment is usually very restricted and might be not compatible with. the browsers (e.g., it is not compatible with Chrome, therefore, you need to use third party projects like @sparticuz/chromium to produce custom Chromium builds).

**Comment by whimboo** [comment]
As reported there is no detail why Firefox doesn't work. What kind of error message is raised by Puppeteer or Firefox? I also would suggest to start Firefox with the following preference set to get trace level output:

```javascript
    extraPrefsFirefox: {
      "remote.log.level": "Trace",
    },
```

**Comment by telltokichu** [comment]
@whimboo
this are the Errors we encounter when trying to run puppeteer with firefox dockerized in aws lambda environment
- Could not find Firefox (rev. latest). This can occur if either\n 1. you did not perform an installation for Firefox before running the script (e.g. `npx puppeteer browsers install firefox`) or\n 2. your cache path is incorrectly configured.
- Failed to launch the browser process!\n[0912/071223.020242:ERROR:headless_shell.cc(266)] Multiple targets are not supported.\n\n\n

Below is the docker file :
![Screenshot 2024-09-16 154534](https://github.com/user-attachments/assets/2a06fc27-275f-4d91-a1c6-ab44ce5145ac)

**Comment by whimboo** [comment]
Is there a reason why you download and install Firefox yourself before running `npm i` for Firefox and even `npx puppeteer browsers install firefox`? Are you sure that Puppeteer has the correct path to the Firefox executable? How does test look like?

Also please still apply the preference from my last comment and provide the logs. If Firefox isn't launched at all it should be a Puppeteer issue.

**Comment by whimboo** [comment]
The first thing would be to make sure to run the right Firefox version. If you are able to start it then really add the `remote.log.level` preference and especially set `DUMPIO=1` so that all the logging output from Puppeteer actually reaches the console.

**Comment by ajanthaselvam** [comment]
Hello, 
 i tried to deploy puppeteer with firefox in lambda, it works fine locally. But while testing in lamdba it  throws error 
(firefox:20): dconf-CRITICAL **: 13:41:21.966: unable to create directory '/home/sbx_user1051/.cache/dconf

**Comment by whimboo** [comment]
Reopening so that we can investigate why it's failing in AWS.

**Comment by whimboo** [comment]
Did you all actually add `browser: "firefox"` to the launch options? If not this could cause this failure.

If that's the case this issue is a dupe of #13113.

**Comment by ajanthaselvam** [comment]
Yes, i have added browser: "firefox".
It works in ECS and not in Lambda. 
As Lambda file system is read only , tried to add firefox in EFS.
const firefoxPath =
      "/mnt/efs/firefox-binary/.cache/puppeteer/firefox/linux-stable_130.0.1/firefox/firefox";
 const browser = await puppeteer.launch({
      browser: "firefox",
      headless: "new",
      protocol: "webDriverBiDi",
      timeout: 0,
      args: [
        "--no-sandbox",
        "--disable-setuid-sandbox",
        "--disable-gpu", // Disable GPU usage
        "--no-zygote",
        "--single-process",
      ],
      executablePath: firefoxPath,
    });

**Comment by whimboo** [comment]
If the Lambda file system is read-only where could Firefox actually write profile data? Is there a temporary path that can be used as "session data"? Would the `/tmp` folder actually work here?

**Comment by ajanthaselvam** [comment]
yes tried /tmp, unfortunately it did not work.
So we mounted EFS to lambda & added the path in executablePath.

**Comment by whimboo** [comment]
The Firefox profile location depends on the `temporaryDirectory` setting for Puppeteer that can be controlled via the configuration or the `PUPPETEER_TMP_DIR` environment variable. Could you try to run your code by setting this environment variable to a path that is actually writeable by the user? Maybe the default config doesn't use the normal temporary folder so it's not writeable?

**Comment by ajanthaselvam** [comment]
Thanks Whimboo! It worked finally.
Added Lambda env variable Home=/tmp & increased ephemeral memory .
will paste the cleaner code asap.

**Comment by whimboo** [comment]
That's great to hear that it's just a write permission issue.

@OrKoN, doesn't Puppeteer automatically use `/tmp` for creating the Firefox profile? I'm not sure why the `HOME` environment variable needs to be modified.

**Comment by OrKoN** [comment]
Puppeteer creates a tmp dir using https://nodejs.org/docs/latest/api/fs.html#fspromisesmkdtempprefix-options and https://nodejs.org/docs/latest/api/os.html#ostmpdir if custom tmp directory is not provided. I do not think these require a HOME variable but it might be that TMPDIR, TMP or TEMP env variables are set to HOME.

**Comment by whimboo** [comment]
Can anyone using AWS lambda please check the values of the above pointed out environment variables? Thanks.

**Comment by QAnders** [comment]
I can't seem to find that `PUPPETEER_TMP_DIR` is being used at all. There's no code (that I can find) that loads `PUPPETEER_TMP_DIR` from env.var and setting `PUPPETEER_TMP_DIR` as a Lambda Environment variable has no effect and `/tmp` is being used.

I also opened an issue here:
#https://github.com/Sparticuz/chromium/issues/231

> I assume it's more of a Puppeteer issue though, but we have the same issue when trying to generate a large number of PDF's... We get a few thousand Invoices (as XML) from a customer and then we need to generate PDF's for these. We have set ephemeral storage to 10GB and we can generate some 600-700 PDF's (invokations) before it runs out of /tmp space, with either ENOSPC: no space left on device, mkdtemp '/tmp/ or
```
2024-10-16T11:55:09.278Z	c5b6c5f7-d891-4760-a2a7-0c2659ab20a8	INFO	htmltopdf getting this error Error: Failed to launch the browser process!
[1016/115509.269172:ERROR:platform_thread_posix.cc(155)] pthread_create: Resource temporarily unavailable (11)
[1016/115509.269211:ERROR:platform_thread_posix.cc(155)] pthread_create: Resource temporarily unavailable (11)
[1016/115509.269328:ERROR:platform_thread_posix.cc(155)] pthread_create: Resource temporarily unavailable (11)
[1016/115509.269544:FATAL:browser_task_executor.cc(310)] Failed to start BrowserThread:IO
````
> AWS is not using "isolated" /tmp storage per invokation, even if we set ephemeral storage size for a Lambda, so it runs out of /tmp space when we have concurrent invokations. We can't delete /tmp in one Lambda either, as each concurrent Lambda is using the same "disk".
> We tried this and then the other concurrent Lambdas "crash" as the /tmp files they had written is gone.

> So, is it possible to know which files "belongs" to which invokation to only delete the ones belonging to that instance?

**Comment by Fanarito** [comment]
Took a lot of trial and error but I finally got firefox and puppeteer to play nicely inside the AWS lambda environment.

Here's a gist with a dockerfile and the launch parameters I used for puppeteer: https://gist.github.com/Fanarito/b1c953c482bb09001c53b7d505844e03

Firefox essentially hung when $HOME was set to a readonly filesystem so setting it to the /tmp directory seemed to solve that.

**Comment by QAnders** [comment]
Nice @Fanarito! We are running a native Lambda so I have been trying to figure out how to set a fixed /tmp/{something unique} path so I can delete the directory once Puppeteer is done.

AWS ephemeral storage is total BS as it is shared between invokations. AWS claims it's designed this way for it to be able to be used as a cache between Lambda invokations running "hot".
